include/oneapi/dpl/pstl/algorithm_ranges_impl.h

```diff
@@ -0,0 +1,313 @@
+{
+    return __pattern_for_each_impl(__tag, std::forward<_ExecutionPolicy>(__exec), std::forward<_R>(__r), __f, __proj);
+}
+
+template <typename _Tag, typename _ExecutionPolicy, typename _R, typename _Proj, typename _Fun>
+auto
+__pattern_for_each(_Tag __tag, _ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj)
+{
+    if constexpr(typename _Tag::__is_vector{})
+        return __pattern_for_each_impl(__tag, std::forward<_ExecutionPolicy>(__exec), std::forward<_R>(__r), __f, __proj);
+    else
+        return std::ranges::for_each(std::forward<_R>(__r), __f, __proj);
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_transform
+//---------------------------------------------------------------------------------------------------------------------
+
+template<typename _Tag, typename _ExecutionPolicy, typename _InRange, typename _OutRange, typename _F, typename _Proj>
+auto
+__pattern_transform_impl(_Tag __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r,
+                    _F __op, _Proj __proj)
+{
+    static_assert(__is_parallel_tag_v<_Tag> || typename _Tag::__is_vector{});
+    assert(__in_r.size() == __out_r.size());
+
+    auto __unary_op = [=](auto&& __val) -> decltype(auto) { return __op(__proj(__val));};
+
+    oneapi::dpl::__internal::__pattern_walk2(__tag, std::forward<_ExecutionPolicy>(__exec), std::ranges::begin(__in_r),
+        std::ranges::begin(__in_r) + __in_r.size(), std::ranges::begin(__out_r),
+        oneapi::dpl::__internal::__transform_functor<decltype(__unary_op)>{std::move(__unary_op)});
+
+    using __return_t = std::ranges::unary_transform_result<std::ranges::borrowed_iterator_t<_InRange>,
+        std::ranges::borrowed_iterator_t<_OutRange>>;
+
+    return __return_t{__in_r.begin() + __in_r.size(), __out_r.begin() + __out_r.size()};
+}
+
+template<typename _IsVector, typename _ExecutionPolicy, typename _InRange, typename _OutRange, typename _F,
+         typename _Proj>
+auto
+__pattern_transform(__parallel_tag<_IsVector> __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r,
+                    _F __op, _Proj __proj)
+{
+    return __pattern_transform_impl(__tag, std::forward<_ExecutionPolicy>(__exec), std::forward<_InRange>(__in_r),
+                                    std::forward<_OutRange>(__out_r), __op, __proj);
+}
+
+template<typename _Tag, typename _ExecutionPolicy, typename _InRange, typename _OutRange, typename _F, typename _Proj>
+auto
+__pattern_transform(_Tag __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r,
+                    _F __op, _Proj __proj)
+{
+    if constexpr(typename _Tag::__is_vector{})
+        return __pattern_transform_impl(__tag, std::forward<_ExecutionPolicy>(__exec), std::forward<_InRange>(__in_r),
+                                        std::forward<_OutRange>(__out_r), __op, __proj);
```

Comment:
We agreed to have the reasonable differences from C++ standard ranges API for serial algorithms. One of them is to use Output range instead of output iterator. 
The reasons:
"
1. It creates a safer API where all the data sequences have known limits.
2. More importantly, not for all algorithms the output size can be calculated based on the input size. An example is copy_if (and similar algorithms with filtering semantics), where the output sequence is allowed to be shorter than the input. Knowing the expected size of the output opens opportunities for more efficient parallel implementations. 
"
(see https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p3179r1.html)
 
And there is third reason - it is impossible to write `__out_r.begin()` for sycl::buffer or sycl::accessor or onedpl::views:all(buffer) on the host side.

Reply:
Thanks, that makes sense.